### PR TITLE
Handle empty package.json files

### LINKED ISSUE #39923: TS Server fatal error:  Maximum call stack size exceeded
Issue Type: <b>Bug</b>

❗️❗️❗️ Please fill in the sections below to help us diagnose the issue ❗️❗️❗️

**TypeScript Version:** 4.1.0-dev.20200804

**Steps to reproduce crash**
1. I created a js file called main.js and it just crashed over and over again.
2.
3.

**TS Server Log**
❗️Server logging disabled. To help us fix crashes like this, please enable logging by setting:
```json
"typescript.tsserver.log": "verbose"
```
After enabling this setting, future crash reports will include the server log.

**TS Server Error Stack**
Server: `semantic`
```
RangeError: Maximum call stack size exceeded
    at toFileNameLowerCase (c:\Users\dmitr\.vscode-insiders\extensions\ms-vscode.vscode-typescript-next-4.1.20200804\node_modules\typescript\lib\tsserver.js:1555:40)
    at Object.toPath (c:\Users\dmitr\.vscode-insiders\extensions\ms-vscode.vscode-typescript-next-4.1.20200804\node_modules\typescript\lib\tsserver.js:6709:16)
    at ProjectService.toPath (c:\Users\dmitr\.vscode-insiders\extensions\ms-vscode.vscode-typescript-next-4.1.20200804\node_modules\typescript\lib\tsserver.js:149798:27)
    at c:\Users\dmitr\.vscode-insiders\extensions\ms-vscode.vscode-typescript-next-4.1.20200804\node_modules\typescript\lib\tsserver.js:152565:56
    at Object.forEachAncestorDirectory (c:\Users\dmitr\.vscode-insiders\extensions\ms-vscode.vscode-typescript-next-4.1.20200804\node_modules\typescript\lib\tsserver.js:6913:26)
    at Object.searchDirectoryAndAncestors (c:\Users\dmitr\.vscode-insiders\extensions\ms-vscode.vscode-typescript-next-4.1.20200804\node_modules\typescript\lib\tsserver.js:152561:24)
    at processDirectory (c:\Users\dmitr\.vscode-insiders\extensions\ms-vscode.vscode-typescript-next-4.1.20200804\node_modules\typescript\lib\tsserver.js:152458:46)
    at processDirectory (c:\Users\dmitr\.vscode-insiders\extensions\ms-vscode.vscode-typescript-next-4.1.20200804\node_modules\typescript\lib\tsserver.js:152459:36)
    at processDirectory (c:\Users\dmitr\.vscode-insiders\extensions\ms-vscode.vscode-typescript-next-4.1.20200804\node_modules\typescript\lib\tsserver.js:152459:36)
    at processDirectory (c:\Users\dmitr\.vscode-insiders\extensions\ms-vscode.vscode-typescript-next-4.1.20200804\node_modules\typescript\lib\tsserver.js:152459:36)
    at processDirectory (c:\Users\dmitr\.vscode-insiders\extensions\ms-vscode.vscode-typescript-next-4.1.20200804\node_modules\typescript\lib\tsserver.js:152459:36)
    at processDirectory (c:\Users\dmitr\.vscode-insiders\extensions\ms-vscode.vscode-typescript-next-4.1.20200804\node_modules\typescript\lib\tsserver.js:152459:36)
    at processDirectory (c:\Users\dmitr\.vscode-insiders\extensions\ms-vscode.vscode-typescript-next-4.1.20200804\node_modules\typescript\lib\tsserver.js:152459:36)
    at processDirectory (c:\Users\dmitr\.vscode-insiders\extensions\ms-vscode.vscode-typescript-next-4.1.20200804\node_modules\typescript\lib\tsserver.js:152459:36)
    at processDirectory (c:\Users\dmitr\.vscode-insiders\extensions\ms-vscode.vscode-typescript-next-4.1.20200804\node_modules\typescript\lib\tsserver.js:152459:36)
    at processDirectory (c:\Users\dmitr\.vscode-insiders\extensions\ms-vscode.vscode-typescript-next-4.1.20200804\node_modules\typescript\lib\tsserver.js:152459:36)
    at processDirectory (c:\Users\dmitr\.vscode-insiders\extensions\ms-vscode.vscode-typescript-next-4.1.20200804\node_modules\typescript\lib\tsserver.js:152459:36)
    at processDirectory (c:\Users\dmitr\.vscode-insiders\extensions\ms-vscode.vscode-typescript-next-4.1.20200804\node_modules\typescript\lib\tsserver.js:152459:36)
    at processDirectory (c:\Users\dmitr\.vscode-insiders\extensions\ms-vscode.vscode-typescript-next-4.1.20200804\node_modules\typescript\lib\tsserver.js:152459:36)
    at processDirectory (c:\Users\dmitr\.vscode-insiders\extensions\ms-vscode.vscode-typescript-next-4.1.20200804\node_modules\typescript\lib\tsserver.js:152459:36)
    at processDirectory (c:\Users\dmitr\.vscode-insiders\extensions\ms-vscode.vscode-typescript-next-4.1.20200804\node_modules\typescript\lib\tsserver.js:152459:36)
    at processDirectory (c:\Users\dmitr\.vscode-insiders\extensions\ms-vscode.vscode-typescript-next-4.1.20200804\node_modules\typescript\lib\tsserver.js:152459:36)
    at processDirectory (c:\Users\dmitr\.vscode-insiders\extensions\ms-vscode.vscode-typescript-next-4.1.20200804\node_modules\typescript\lib\tsserver.js:152459:36)
    at processDirectory (c:\Users\dmitr\.vscode-insiders\extensions\ms-vscode.vscode-typescript-next-4.1.20200804\node_modules\typescript\lib\tsserver.js:152459:36)
    at processDirectory (c:\Users\dmitr\.vscode-insiders\extensions\ms-vscode.vscode-typescript-next-4.1.20200804\node_modules\typescript\lib\tsserver.js:152459:36)
    at processDirectory (c:\Users\dmitr\.vscode-insiders\extensions\ms-vscode.vscode-typescript-next-4.1.20200804\node_modules\typescript\lib\tsserver.js:152459:36)
    at processDirectory (c:\Users\dmitr\.vscode-insiders\extensions\ms-vscode.vscode-typescript-next-4.1.20200804\node_modules\typescript\lib\tsserver.js:152459:36)
    at processDirectory (c:\Users\dmitr\.vscode-insiders\extensions\ms-vscode.vscode-typescript-next-4.1.20200804\node_modules\typescript\lib\tsserver.js:152459:36)
    at processDirectory (c:\Users\dmitr\.vscode-insiders\extensions\ms-vscode.vscode-typescript-next-4.1.20200804\node_modules\typescript\lib\tsserver.js:152459:36)
    at processDirectory (c:\Users\dmitr\.vscode-insiders\extensions\ms-vscode.vscode-typescript-next-4.1.20200804\node_modules\typescript\lib\tsserver.js:152459:36)
    at processDirectory (c:\Users\dmitr\.vscode-insiders\extensions\ms-vscode.vscode-typescript-next-4.1.20200804\node_modules\typescript\lib\tsserver.js:152459:36)
    at processDirectory (c:\Users\dmitr\.vscode-insiders\extensions\ms-vscode.vscode-typescript-next-4.1.20200804\node_modules\typescript\lib\tsserver.js:152459:36)
    at processDirectory (c:\Users\dmitr\.vscode-insiders\extensions\ms-vscode.vscode-typescript-next-4.1.20200804\node_modules\typescript\lib\tsserver.js:152459:36)
    at processDirectory (c:\Users\dmitr\.vscode-insiders\extensions\ms-vscode.vscode-typescript-next-4.1.20200804\node_modules\typescript\lib\tsserver.js:152459:36)
    at processDirectory (c:\Users\dmitr\.vscode-insiders\extensions\ms-vscode.vscode-typescript-next-4.1.20200804\node_modules\typescript\lib\tsserver.js:152459:36)
    at processDirectory (c:\Users\dmitr\.vscode-insiders\extensions\ms-vscode.vscode-typescript-next-4.1.20200804\node_modules\typescript\lib\tsserver.js:152459:36)
    at processDirectory (c:\Users\dmitr\.vscode-insiders\extensions\ms-vscode.vscode-typescript-next-4.1.20200804\node_modules\typescript\lib\tsserver.js:152459:36)
    at processDirectory (c:\Users\dmitr\.vscode-insiders\extensions\ms-vscode.vscode-typescript-next-4.1.20200804\node_modules\typescript\lib\tsserver.js:152459:36)
    at processDirectory (c:\Users\dmitr\.vscode-insiders\extensions\ms-vscode.vscode-typescript-next-4.1.20200804\node_modules\typescript\lib\tsserver.js:152459:36)
    at processDirectory (c:\Users\dmitr\.vscode-insiders\extensions\ms-vscode.vscode-typescript-next-4.1.20200804\node_modules\typescript\lib\tsserver.js:152459:36)
    at processDirectory (c:\Users\dmitr\.vscode-insiders\extensions\ms-vscode.vscode-typescript-next-4.1.20200804\node_modules\typescript\lib\tsserver.js:152459:36)
    at processDirectory (c:\Users\dmitr\.vscode-insiders\extensions\ms-vscode.vscode-typescript-next-4.1.20200804\node_modules\typescript\lib\tsserver.js:152459:36)
    at processDirectory (c:\Users\dmitr\.vscode-insiders\extensions\ms-vscode.vscode-typescript-next-4.1.20200804\node_modules\typescript\lib\tsserver.js:152459:36)
    at processDirectory (c:\Users\dmitr\.vscode-insiders\extensions\ms-vscode.vscode-typescript-next-4.1.20200804\node_modules\typescript\lib\tsserver.js:152459:36)
    at processDirectory (c:\Users\dmitr\.vscode-insiders\extensions\ms-vscode.vscode-typescript-next-4.1.20200804\node_modules\typescript\lib\tsserver.js:152459:36)
    at processDirectory (c:\Users\dmitr\.vscode-insiders\extensions\ms-vscode.vscode-typescript-next-4.1.20200804\node_modules\typescript\lib\tsserver.js:152459:36)
    at processDirectory (c:\Users\dmitr\.vscode-insiders\extensions\ms-vscode.vscode-typescript-next-4.1.20200804\node_modules\typescript\lib\tsserver.js:152459:36)
    at processDirectory (c:\Users\dmitr\.vscode-insiders\extensions\ms-vscode.vscode-typescript-next-4.1.20200804\node_modules\typescript\lib\tsserver.js:152459:36)
    at processDirectory (c:\Users\dmitr\.vscode-insiders\extensions\ms-vscode.vscode-typescript-next-4.1.20200804\node_modules\typescript\lib\tsserver.js:152459:36)
    at processDirectory (c:\Users\dmitr\.vscode-insiders\extensions\ms-vscode.vscode-typescript-next-4.1.20200804\node_modules\typescript\lib\tsserver.js:152459:36)
    at processDirectory (c:\Users\dmitr\.vscode-insiders\extensions\ms-vscode.vscode-typescript-next-4.1.20200804\node_modules\typescript\lib\tsserver.js:152459:36)
    at processDirectory (c:\Users\dmitr\.vscode-insiders\extensions\ms-vscode.vscode-typescript-next-4.1.20200804\node_modules\typescript\lib\tsserver.js:152459:36)
    at processDirectory (c:\Users\dmitr\.vscode-insiders\extensions\ms-vscode.vscode-typescript-next-4.1.20200804\node_modules\typescript\lib\tsserver.js:152459:36)
    at processDirectory (c:\Users\dmitr\.vscode-insiders\extensions\ms-vscode.vscode-typescript-next-4.1.20200804\node_modules\typescript\lib\tsserver.js:152459:36)
    at processDirectory (c:\Users\dmitr\.vscode-insiders\extensions\ms-vscode.vscode-typescript-next-4.1.20200804\node_modules\typescript\lib\tsserver.js:152459:36)
    at processDirectory (c:\Users\dmitr\.vscode-insiders\extensions\ms-vscode.vscode-typescript-next-4.1.20200804\node_modules\typescript\lib\tsserver.js:152459:36)
    at processDirectory (c:\Users\dmitr\.vscode-insiders\extensions\ms-vscode.vscode-typescript-next-4.1.20200804\node_modules\typescript\lib\tsserver.js:152459:36)
    at processDirectory (c:\Users\dmitr\.vscode-insiders\extensions\ms-vscode.vscode-typescript-next-4.1.20200804\node_modules\typescript\lib\tsserver.js:152459:36)
    at processDirectory (c:\Users\dmitr\.vscode-insiders\extensions\ms-vscode.vscode-typescript-next-4.1.20200804\node_modules\typescript\lib\tsserver.js:152459:36)
    at processDirectory (c:\Users\dmitr\.vscode-insiders\extensions\ms-vscode.vscode-typescript-next-4.1.20200804\node_modules\typescript\lib\tsserver.js:152459:36)
    at processDirectory (c:\Users\dmitr\.vscode-insiders\extensions\ms-vscode.vscode-typescript-next-4.1.20200804\node_modules\typescript\lib\tsserver.js:152459:36)
    at processDirectory (c:\Users\dmitr\.vscode-insiders\extensions\ms-vscode.vscode-typescript-next-4.1.20200804\node_modules\typescript\lib\tsserver.js:152459:36)
    at processDirectory (c:\Users\dmitr\.vscode-insiders\extensions\ms-vscode.vscode-typescript-next-4.1.20200804\node_modules\typescript\lib\tsserver.js:152459:36)
    at processDirectory (c:\Users\dmitr\.vscode-insiders\extensions\ms-vscode.vscode-typescript-next-4.1.20200804\node_modules\typescript\lib\tsserver.js:152459:36)
    at processDirectory (c:\Users\dmitr\.vscode-insiders\extensions\ms-vscode.vscode-typescript-next-4.1.20200804\node_modules\typescript\lib\tsserver.js:152459:36)
    at processDirectory (c:\Users\dmitr\.vscode-insiders\extensions\ms-vscode.vscode-typescript-next-4.1.20200804\node_modules\typescript\lib\tsserver.js:152459:36)
    at processDirectory (c:\Users\dmitr\.vscode-insiders\extensions\ms-vscode.vscode-typescript-next-4.1.20200804\node_modules\typescript\lib\tsserver.js:152459:36)
    at processDirectory (c:\Users\dmitr\.vscode-insiders\extensions\ms-vscode.vscode-typescript-next-4.1.20200804\node_modules\typescript\lib\tsserver.js:152459:36)
    at processDirectory (c:\Users\dmitr\.vscode-insiders\extensions\ms-vscode.vscode-typescript-next-4.1.20200804\node_modules\typescript\lib\tsserver.js:152459:36)
    at processDirectory (c:\Users\dmitr\.vscode-insiders\extensions\ms-vscode.vscode-typescript-next-4.1.20200804\node_modules\typescript\lib\tsserver.js:152459:36)
    at processDirectory (c:\Users\dmitr\.vscode-insiders\extensions\ms-vscode.vscode-typescript-next-4.1.20200804\node_modules\typescript\lib\tsserver.js:152459:36)
    at processDirectory (c:\Users\dmitr\.vscode-insiders\extensions\ms-vscode.vscode-typescript-next-4.1.20200804\node_modules\typescript\lib\tsserver.js:152459:36)
    at processDirectory (c:\Users\dmitr\.vscode-insiders\extensions\ms-vscode.vscode-typescript-next-4.1.20200804\node_modules\typescript\lib\tsserver.js:152459:36)
    at processDirectory (c:\Users\dmitr\.vscode-insiders\extensions\ms-vscode.vscode-typescript-next-4.1.20200804\node_modules\typescript\lib\tsserver.js:152459:36)
    at processDirectory (c:\Users\dmitr\.vscode-insiders\extensions\ms-vscode.vscode-typescript-next-4.1.20200804\node_modules\typescript\lib\tsserver.js:152459:36)
    at processDirectory (c:\Users\dmitr\.vscode-insiders\extensions\ms-vscode.vscode-typescript-next-4.1.20200804\node_modules\typescript\lib\tsserver.js:152459:36)
    at processDirectory (c:\Users\dmitr\.vscode-insiders\extensions\ms-vscode.vscode-typescript-next-4.1.20200804\node_modules\typescript\lib\tsserver.js:152459:36)
    at processDirectory (c:\Users\dmitr\.vscode-insiders\extensions\ms-vscode.vscode-typescript-next-4.1.20200804\node_modules\typescript\lib\tsserver.js:152459:36)
    at processDirectory (c:\Users\dmitr\.vscode-insiders\extensions\ms-vscode.vscode-typescript-next-4.1.20200804\node_modules\typescript\lib\tsserver.js:152459:36)
    at processDirectory (c:\Users\dmitr\.vscode-insiders\extensions\ms-vscode.vscode-typescript-next-4.1.20200804\node_modules\typescript\lib\tsserver.js:152459:36)
    at processDirectory (c:\Users\dmitr\.vscode-insiders\extensions\ms-vscode.vscode-typescript-next-4.1.20200804\node_modules\typescript\lib\tsserver.js:152459:36)
    at processDirectory (c:\Users\dmitr\.vscode-insiders\extensions\ms-vscode.vscode-typescript-next-4.1.20200804\node_modules\typescript\lib\tsserver.js:152459:36)
    at processDirectory (c:\Users\dmitr\.vscode-insiders\extensions\ms-vscode.vscode-typescript-next-4.1.20200804\node_modules\typescript\lib\tsserver.js:152459:36)
    at processDirectory (c:\Users\dmitr\.vscode-insiders\extensions\ms-vscode.vscode-typescript-next-4.1.20200804\node_modules\typescript\lib\tsserver.js:152459:36)
    at processDirectory (c:\Users\dmitr\.vscode-insiders\extensions\ms-vscode.vscode-typescript-next-4.1.20200804\node_modules\typescript\lib\tsserver.js:152459:36)
    at processDirectory (c:\Users\dmitr\.vscode-insiders\extensions\ms-vscode.vscode-typescript-next-4.1.20200804\node_modules\typescript\lib\tsserver.js:152459:36)
    at processDirectory (c:\Users\dmitr\.vscode-insiders\extensions\ms-vscode.vscode-typescript-next-4.1.20200804\node_modules\typescript\lib\tsserver.js:152459:36)
    at processDirectory (c:\Users\dmitr\.vscode-insiders\extensions\ms-vscode.vscode-typescript-next-4.1.20200804\node_modules\typescript\lib\tsserver.js:152459:36)
    at processDirectory (c:\Users\dmitr\.vscode-insiders\extensions\ms-vscode.vscode-typescript-next-4.1.20200804\node_modules\typescript\lib\tsserver.js:152459:36)
    at processDirectory (c:\Users\dmitr\.vscode-insiders\extensions\ms-vscode.vscode-typescript-next-4.1.20200804\node_modules\typescript\lib\tsserver.js:152459:36)
    at processDirectory (c:\Users\dmitr\.vscode-insiders\extensions\ms-vscode.vscode-typescript-next-4.1.20200804\node_modules\typescript\lib\tsserver.js:152459:36)
    at processDirectory (c:\Users\dmitr\.vscode-insiders\extensions\ms-vscode.vscode-typescript-next-4.1.20200804\node_modules\typescript\lib\tsserver.js:152459:36)
    at processDirectory (c:\Users\dmitr\.vscode-insiders\extensions\ms-vscode.vscode-typescript-next-4.1.20200804\node_modules\typescript\lib\tsserver.js:152459:36)
    at processDirectory (c:\Users\dmitr\.vscode-insiders\extensions\ms-vscode.vscode-typescript-next-4.1.20200804\node_modules\typescript\lib\tsserver.js:152459:36)
    at processDirectory (c:\Users\dmitr\.vscode-insiders\extensions\ms-vscode.vscode-typescript-next-4.1.20200804\node_modules\typescript\lib\tsserver.js:152459:36)
    at processDirectory (c:\Users\dmitr\.vscode-insiders\extensions\ms-vscode.vscode-typescript-next-4.1.20200804\node_modules\typescript\lib\tsserver.js:152459:36)
    at processDirectory (c:\Users\dmitr\.vscode-insiders\extensions\ms-vscode.vscode-typescript-next-4.1.20200804\node_modules\typescript\lib\tsserver.js:152459:36)
    at processDirectory (c:\Users\dmitr\.vscode-insiders\extensions\ms-vscode.vscode-typescript-next-4.1.20200804\node_modules\typescript\lib\tsserver.js:152459:36)
    at processDirectory (c:\Users\dmitr\.vscode-insiders\extensions\ms-vscode.vscode-typescript-next-4.1.20200804\node_modules\typescript\lib\tsserver.js:152459:36)
    at processDirectory (c:\Users\dmitr\.vscode-insiders\extensions\ms-vscode.vscode-typescript-next-4.1.20200804\node_modules\typescript\lib\tsserver.js:152459:36)
```

VS Code version: Code - Insiders 1.48.0-insider (d13d2fc56da7a2f8bcad4256212db0661fcbba45, 2020-08-05T05:30:25.429Z)
OS version: Windows_NT x64 10.0.18363

<details>
<summary>System Info</summary>

|Item|Value|
|---|---|
|CPUs|AMD Ryzen 5 3600 6-Core Processor               (12 x 3593)|
|GPU Status|2d_canvas: enabled<br>flash_3d: enabled<br>flash_stage3d: enabled<br>flash_stage3d_baseline: enabled<br>gpu_compositing: enabled<br>multiple_raster_threads: enabled_on<br>oop_rasterization: disabled_off<br>protected_video_decode: unavailable_off<br>rasterization: enabled<br>skia_renderer: disabled_off_ok<br>video_decode: enabled<br>viz_display_compositor: enabled_on<br>viz_hit_test_surface_layer: disabled_off_ok<br>webgl: enabled<br>webgl2: enabled|
|Load (avg)|undefined|
|Memory (System)|31.93GB (20.08GB free)|
|Process Argv||
|Screen Reader|no|
|VM|0%|
</details><details><summary>Extensions (7)</summary>

Extension|Author (truncated)|Version
---|---|---
vscode-pull-request-github|Git|0.18.1
discord-vscode|icr|4.0.0
csharp|ms-|1.22.1
python|ms-|2020.7.96456
cpptools|ms-|0.29.0
vscode-typescript-next|ms-|4.1.20200804
vscodeintellicode|Vis|1.2.9

(1 theme extensions excluded)

</details>

### PR DESCRIPTION
**Pull request by andrewbranch**
Fixes #39923

Thanks for investigating @DanielRosenwasser; it was easier for me just to throw this together since I knew where to put a test.

### CODE REVIEW

File: src/services/utilities.ts [hunk 1]
```diff
@@ -2735,9 +2735,7 @@ namespace ts {
 
         type PackageJsonRaw = Record<typeof dependencyKeys[number], Record<string, string> | undefined>;
         const dependencyKeys = ["dependencies", "devDependencies", "optionalDependencies", "peerDependencies"] as const;
-        const stringContent = host.readFile(fileName);
-        if (!stringContent) return undefined;
```
Comment (andrewbranch) [inline src/services/utilities.ts:2739]: This was the problem. Non-empty package.jsons with parse errors were already handled, but empty files were not.

**Comment by DanielRosenwasser** [comment]
@typescript-bot pack this

**Comment by DanielRosenwasser** [comment]
@typescript-bot cherry-pick this to release-4.0

File: src/server/editorServices.ts [hunk 2]
```diff
@@ -3774,12 +3774,10 @@ namespace ts.server {
...
             const packageJsonCache = this.packageJsonCache;
-            const watchPackageJsonFile = this.watchPackageJsonFile.bind(this);
-            const toPath = this.toPath.bind(this);
-            const rootPath = rootDir && toPath(rootDir);
-            const filePath = toPath(fileName);
+            const rootPath = rootDir && this.toPath(rootDir);
+            const filePath = this.toPath(fileName);
             const result: PackageJsonInfo[] = [];
-            forEachAncestorDirectory(getDirectoryPath(filePath), function processDirectory(directory): boolean | undefined {
+            const processDirectory = (directory: Path): boolean | undefined => {
```
Comment (andrewbranch) [inline src/server/editorServices.ts:3780]: All the changes in this file are just to eliminate the `.bind(this)` which you noticed were `any`.

**Comment by DanielRosenwasser** [comment]
@typescript-bot cherry-pick this to release-4.0 and LKG

**Comment by DanielRosenwasser** [comment]
Confirmed, no longer crashes for me.

**Pull request merged**